controllers/infra/controllers/infraresource_controller.go

```diff
@@ -0,0 +1,222 @@
+		if err = r.Status().Update(ctx, infraController); err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		r.Logger.Error(err, "UpdateResourceUsed failed")
+		return fmt.Errorf("infra controller update err:%v", err)
+	}
+	r.Logger.Info("infra controller calculate resource success")
+	return nil
+}
+
+// checkInfraStatus check infra status
+func (r *InfraResourceReconciler) checkInfraStatus(infra infrav1.Infra) bool {
+	if infra.Status.Status == infrav1.Running.String() {
+		return true
+	}
+	r.Logger.Info("infra status is not running", "infra name", infra.Name, "infra namespace", infra.Namespace, "infra status", infra.Status.Status)
+	return false
+}
+
+// syncInfraResource sync resource used for each infra
+func (r *InfraResourceReconciler) syncInfraResource(ctx context.Context, infra infrav1.Infra, infraController *infrav1.InfraResource) error {
+	var cpuResourceQuantity int64
+	var memoryResourceQuantity int64
+	var volumeResourceQuantity int64
+	for _, host := range infra.Spec.Hosts {
+		cnt := int64(host.Count)
+		flavor := host.Flavor
+		cpuResourceQuantity += common.CPUMap[flavor] * cnt
+		memoryResourceQuantity += common.MemoryMap[flavor] * cnt
+		for _, disk := range host.Disks {
+			volumeResourceQuantity += int64(disk.Capacity)
+		}
+		volumeResourceQuantity = volumeResourceQuantity * int64(host.Count)
+	}
+	quantityMap := make(map[v1.ResourceName]int64)
+	quantityMap[common.CPUResourceName] = cpuResourceQuantity
+	quantityMap[common.MemoryResourceName] = memoryResourceQuantity
+	quantityMap[common.VolumeResourceName] = volumeResourceQuantity
+	infraResource := meteringv1.Resource{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GetResourceName(&infra, infraController.Status.SeqID),
+			Namespace: common.InfraSystemNamespace,
+		},
+	}
+	if _, err := controllerutil.CreateOrUpdate(ctx, r.Client, &infraResource, func() error {
+		if infraResource.Spec.Resources == nil {
+			infraResource.Spec.Resources = make(map[v1.ResourceName]meteringv1.ResourceInfo)
+		}
+		//todo: require the compatible resource CRD
+		//for k, v := range infraController.Spec.Resources {
+		//	infraResource.Spec.Resources[k] = meteringv1.ResourceInfo{
+		//		Used:      quantityMap[k],
+		//		TimeStamp: time.Now().Unix(),
+		//		NameSpace: infra.Namespace,
+		//		Cost:      quantityMap[k] * v.Price,
+		//	}
+		//}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetResourceName get resource name
+func GetResourceName(infra *infrav1.Infra, seqID int64) string {
+	return fmt.Sprintf("%s-%s-%v", infra.Namespace, infra.Name, seqID)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *InfraResourceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	const controllerName = "InfraResource-controller"
+	r.Logger = ctrl.Log.WithName(controllerName)
+	r.Logger.V(1).Info("reconcile InfraResource-controller")
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&infrav1.InfraResource{}).
```

Comment:
can use this to ignore status update lead to reconcile.
```
For(&meteringv1.InfraResource{}, builder.WithPredicates(predicate.Or(predicate.GenerationChangedPredicate{}))).
```

Reply:
> can use this to ignore status update lead to reconcile.
> 
> ```
> For(&meteringv1.InfraResource{}, builder.WithPredicates(predicate.Or(predicate.GenerationChangedPredicate{}))).
> ```

done